sway/desktop/fx_renderer.c

```diff
@@ -371,27 +572,103 @@ struct fx_renderer *fx_renderer_create(struct wlr_egl *egl) {
-	}
-	glBindRenderbuffer(GL_RENDERBUFFER, renderer->stencil_buffer_id);
-	glRenderbufferStorage(GL_RENDERBUFFER, GL_STENCIL_INDEX8, width, height);
-	glFramebufferRenderbuffer(GL_FRAMEBUFFER, GL_STENCIL_ATTACHMENT,
-			GL_RENDERBUFFER, renderer->stencil_buffer_id);
+void fx_renderer_begin(struct fx_renderer *renderer, struct sway_output *sway_output,
+		pixman_region32_t *original_damage) {
+	struct wlr_output *output = sway_output->wlr_output;
 
-	glViewport(0, 0, width, height);
+	uint32_t width = output->width;
+	uint32_t height = output->height;
+
+	renderer->sway_output = sway_output;
+	renderer->original_damage = original_damage;
+	// Store the wlr framebuffer
+	glGetIntegerv(GL_FRAMEBUFFER_BINDING, &renderer->wlr_fb);
+
+	// Create the main framebuffer
+	create_fx_framebuffer(output, &renderer->main_buffer, &renderer->main_buffer);
+
+	// Create a new blur/effects framebuffers
+	create_fx_framebuffer(output, &renderer->blur_buffer, &renderer->main_buffer);
+	create_fx_framebuffer(output, &renderer->effects_buffer, &renderer->main_buffer);
+	create_fx_framebuffer(output, &renderer->effects_buffer_swapped, &renderer->main_buffer);
+
+	// Create and render the stencil buffer
+	create_stencil_buffer(output, &renderer->stencil_buffer_id);
 
 	// refresh projection matrix
 	matrix_projection(renderer->projection, width, height,
 		WL_OUTPUT_TRANSFORM_FLIPPED_180);
 
 	glBlendFunc(GL_ONE, GL_ONE_MINUS_SRC_ALPHA);
+
+	// Bind to our main framebuffer
+	bind_framebuffer(renderer->main_buffer, width, height);
 }
 
-void fx_renderer_end() {
-	// TODO
+void render_whole_output(struct fx_renderer *renderer) {
+	struct wlr_output *output = renderer->sway_output->wlr_output;
+	int width, height;
+	wlr_output_transformed_resolution(output, &width, &height);
+
+	struct wlr_box monitor_box = {0, 0, width, height};
+	scale_box(&monitor_box, output->scale);
+
+	float matrix[9];
+	enum wl_output_transform transform =
+		wlr_output_transform_invert(output->transform);
+	wlr_matrix_project_box(matrix, &monitor_box, transform, 0.0,
+		output->transform_matrix);
+
+	pixman_region32_t damage;
+	pixman_region32_init(&damage);
+	pixman_region32_union_rect(&damage, &damage, monitor_box.x, monitor_box.y,
+		monitor_box.width, monitor_box.height);
+	pixman_region32_intersect(&damage, &damage, renderer->original_damage);
+	bool damaged = pixman_region32_not_empty(&damage);
+	if (!damaged) {
+		goto damage_finish;
+	}
+
+	struct decoration_data deco_data = {
+		.alpha = 1.0f,
+		.dim = 0.0f,
+		.dim_color = config->dim_inactive_colors.unfocused,
+		.corner_radius = 0,
+		.saturation = 1.0f,
+		.has_titlebar = false,
+		.blur = false,
+	};
+
+	int nrects;
+	pixman_box32_t *rects = pixman_region32_rectangles(&damage, &nrects);
+	for (int i = 0; i < nrects; ++i) {
+		scissor_output(output, &rects[i]);
+		fx_render_texture_with_matrix(renderer, &renderer->main_buffer.texture, &monitor_box, matrix, deco_data);
+	}
+
+damage_finish:
+	pixman_region32_fini(&damage);
+}
```

Comment:
As part of keeping damage tracking in render.c, this can probably go there. We want to treat fx_renderer.c as more of an opengl wrapper